src/sinks/util/auto_concurrency/controller.rs

```diff
@@ -0,0 +1,282 @@
+#[derive(Clone, Debug)]
+pub(super) struct Controller<L> {
+    semaphore: Arc<ShrinkableSemaphore>,
+    max: usize,
+    logic: L,
+    pub(super) inner: Arc<Mutex<Inner>>,
+    #[cfg(test)]
+    pub(super) stats: Arc<Mutex<ControllerStatistics>>,
+}
+
+#[derive(Debug)]
+pub(super) struct Inner {
+    pub(super) current_limit: usize,
+    in_flight: usize,
+    past_rtt: EWMA,
+    next_update: Instant,
+    current_rtt: Mean,
+    had_back_pressure: bool,
+    reached_limit: bool,
+}
+
+#[cfg(test)]
+#[derive(Debug, Default)]
+pub(super) struct ControllerStatistics {
+    pub(super) in_flight: TimeHistogram,
+    pub(super) concurrency_limit: TimeHistogram,
+    pub(super) observed_rtt: TimeWeightedSum,
+    pub(super) averaged_rtt: TimeWeightedSum,
+}
+
+impl<L> Controller<L> {
+    pub(super) fn new(in_flight_limit: Option<usize>, logic: L) -> Self {
+        // If an in_flight_limit is specified, it becomse both the
+        // current limit and the maximum, effectively bypassing all the
+        // mechanisms. Otherwise, the current limit is set to 1 and the
+        // maximum to MAX_CONCURRENCY.
+        let current_limit = in_flight_limit.unwrap_or(1);
+        let max = in_flight_limit.unwrap_or(super::MAX_CONCURRENCY);
+        Self {
+            semaphore: Arc::new(ShrinkableSemaphore::new(current_limit)),
+            max,
+            logic,
+            inner: Arc::new(Mutex::new(Inner {
+                current_limit,
+                in_flight: 0,
+                past_rtt: Default::default(),
+                next_update: instant_now(),
+                current_rtt: Default::default(),
+                had_back_pressure: false,
+                reached_limit: false,
+            })),
+            #[cfg(test)]
+            stats: Arc::new(Mutex::new(ControllerStatistics::default())),
+        }
+    }
+
+    pub(super) fn acquire(&self) -> impl Future<Output = OwnedSemaphorePermit> + Send + 'static {
+        let mut inner = self.inner.lock().expect("Controller mutex is poisoned");
+        if inner.in_flight >= inner.current_limit {
+            inner.reached_limit = true;
+        }
+        Arc::clone(&self.semaphore).acquire()
+    }
+
+    pub(super) fn start_request(&self) {
+        let mut inner = self.inner.lock().expect("Controller mutex is poisoned");
+
+        #[cfg(test)]
+        {
+            let mut stats = self.stats.lock().expect("Stats mutex is poisoned");
+            stats.in_flight.add(inner.in_flight, instant_now());
+        }
+
+        inner.in_flight += 1;
+        emit!(AutoConcurrencyInFlight {
+            in_flight: inner.in_flight as u64
+        });
+    }
+
+    fn adjust_to_back_pressure(&self, start: Instant, is_back_pressure: bool) {
+        let now = instant_now();
+        let mut inner = self.inner.lock().expect("Controller mutex is poisoned");
+
+        #[cfg(test)]
+        let mut stats = self.stats.lock().expect("Stats mutex is poisoned");
```

Comment:
Might be useful to switch to `try_lock` here if possible - to avoid the effects of additional locking affecting test results.

Reply:
So, if we use `lock`, then the stats may not reflect how the sink will operate in practice because there is additional lock contention. If we use `try_lock`, then the stats may not reflect how the sink actually did operate because it sometimes didn't take the lock and so couldn't report stats, resulting in incomplete data. Either way, stats can't always reflect actual operation, which I agree is unfortunate. Using `try_lock` also requires additional conditionals when trying to report stats. I think I'll stick with `lock` unless there is another consideration.

Now, I can reorganize the code a bit to make the time the mutex is held even shorter, which will reduce the effect. Do you have any thoughts on whether it is better to take the mutex once and hold it for a bit longer in `adjust_to_back_pressure`, or to take it once for each place it is needed and drop it immediately?

Reply:
Yeah, I see the issue with the `try_lock` access to stats.

I had some thoughts on using lock-free ring buffers (channels?) to collect the stats for test purposes, and then to aggregate them separately. But it's very early - I'm still reading through the code and wrapping my head around what's going on.

A bit preliminary, but to me, implementing channel based mocks to tightly control the execution in a tight lock-step looks promising here. This would allow us to assert the states in all the right places, and eliminate the need for manual locking and associated race conditions.

The idea is we use a channel to send notifications that we reached a certain state to the test controlling logic along the way, and we then wait on another rx channel to continue. The test logic waits for a particular notification, does the assertion, and sends a message that lets the tested code continue the execution.

Again, this is very preliminary - I'll have to take a deeper look if this has any actual benefits over what we have already - but I'm very curious to hear what are your thoughts on this.

Reply:
I had considered using channels for reporting test metrics, but couldn't see a path that was as straightforward as using locking. It is worth considering, though.

I see there being roughly 3 levels of testing: low level (does it take the steps that the code says it should), behavior (do the steps give it the expected high-level behavior), and performance (does it interoperate optimally with live servers). It sounds like you are describing some fairly low-level tests, covering some of the steps tested in `auto_concurrency/service.rs` using tokio's timers to manually advance steps. It might be worth covering other paths though.

Reply:
Yes, you're right, the approach I'm thinking about is just briefly above, and very similar to, tracing the code along it's execution in lock-step, and asserting that exactly what's expected happens.

I'm thinking of tokio timers as just one of the ways to advance.
Consider having `state` at the test end rather than as part of the struct, and only passing it into the `Controller` code via a channel. That channel will pass ownership and access to `state`, and it'll have to naturally be passed between the threads to mix the updates and assertions.